### PR TITLE
Implement `yolo test` feature

### PR DESCRIPTION
**Pull request by validatedev**
## Description
This pull request addresses issue https://github.com/ultralytics/ultralytics/issues/1068 in the Ultralytics repository.

The primary focus of this pull request is to implement the feature known as `test`, which aims to simplify the process of utilizing the `test` dataset. To draw a comparison, the functionality of this feature is akin to `task=test` in [YOLOv5](https://github.com/ultralytics/yolov5). This enhancement was initially requested by @glenn-jocher in a comment on [this](https://github.com/ultralytics/ultralytics/pull/704#issuecomment-1409367229) thread.

## Use Case
The proposed `test` feature carries great benefits for users. The key advantage is that it allows users to effortlessly work with their `test` dataset without the necessity of manipulating complex configuration parameters in YOLOv8. This results in a more streamlined and user-friendly interface, especially beneficial to those unfamiliar with the intricate settings of the YOLOv8 platform.

## Implementation Details
In terms of technical execution, the `test` feature acts as a wrapper for `yolo val split=test`. This feature doesn't introduce any complexity since it employs the existing validation implementation, similar to the approach in [YOLOv5](https://github.com/ultralytics/yolov5). It's a matter of reuse and optimization of what is already available rather than creating new, complex solutions. This, in turn, ensures stability and mitigates potential risks associated with new, untested code.

### Summary
🧪🛠️📊

Added a 'test' mode to the YOLO framework to enable model evaluation on different datasets. The `test` mode is implemented as a wrapper for the `val` mode with a different task and split. The `check_det_dataset` and `get_dataloader` functions were modified to support the `split` parameter and the `mode` argument respectively.

> _We test the model on the edge of doom_
> _We split the data and unleash the `YOLO`_
> _We validate the results with `BaseValidator`_
> _We rock the `DetectionValidator` in any mode_

### Walkthrough
*  Add 'test' mode to YOLO framework to enable testing on a given dataset and get metrics ([link](https://github.com/ultralytics/ultralytics/pull/3006/files?diff=unified&w=0#diff-daf1714127b012666c9fed39d0e75184c60be83f0c6bd3f9863f27113b1a6c22L17-R17), [link](https://github.com/ultralytics/ultralytics/pull/3006/files?diff=unified&w=0#diff-daf1714127b012666c9fed39d0e75184c60be83f0c6bd3f9863f27113b1a6c22L387-R387), [link](https://github.com/ultralytics/ultralytics/pull/3006/files?diff=unified&w=0#diff-555e5fdee1244b9d95b48487cfd502d068e5bfcd51001d5f5b3f4a33007bda7cR308-R343), [link](https://github.com/ultralytics/ultralytics/pull/3006/files?diff=unified&w=0#diff-6e518634bff3eacf0c8b7755670115753754e2b4987f0a323bd54a0b83707f02L208-R208))
*  Update `check_det_dataset` function to handle missing 'val' or 'test' split in data dictionary and use alternative split with a warning ([link](https://github.com/ultralytics/ultralytics/pull/3006/files?diff=unified&w=0#diff-15f6dd1ea95b187a5b50c885f567bad50d006f1af1180508497a038247e9df58L193-R193), [link](https://github.com/ultralytics/ultralytics/pull/3006/files?diff=unified&w=0#diff-15f6dd1ea95b187a5b50c885f567bad50d006f1af1180508497a038247e9df58R266-R273), [link](https://github.com/ultralytics/ultralytics/pull/3006/files?diff=unified&w=0#diff-7646fc7a670fe6bbcbf9e637a707139b362bd3b118be7faca28bff7d9b7450c3L127-R127))

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
New Test Mode Documentation and Implementation for Ultralytics YOLOv8 Models

### 📊 Key Changes
- Added a new `Test mode` documentation under `docs/modes/test.md`.
- Updated `docs/quickstart.md` to include instructions for using the new Test mode.
- Included Test mode details in `docs/usage/cfg.md` and `docs/usage/cli.md`.
- Provided usage examples of Test mode in `docs/usage/python.md`.
- Added CLI and Python tests for the Test mode in `tests/test_cli.py` and `tests/test_python.py` respectively.
- Updated `ultralytics/yolo/cfg/__init__.py` to recognize Test mode as a valid argument.
- Adjusted `ultralytics/yolo/data/utils.py` to check and set appropriate dataset paths for testing.
- Expanded `ultralytics/yolo/engine/model.py` and `ultralytics/yolo/engine/validator.py` to handle model testing.
- Refactored Test settings to be aligned with Validation settings for consistency and reuse.

### 🎯 Purpose & Impact
- 🎯 Enables users to easily test trained YOLOv8 models on test datasets to measure model accuracy and generalization performance without manually retaining dataset and settings.
- 🔍 Helps to ensure models perform well on unseen data and can be a routine part of the model development workflow.
- 💡 Provides a seamless experience for users transitioning from training or validating to testing their models, fostering an end-to-end development process.
- 🔧 Offers developers and users a hands-off approach by remembering training settings, allowing simple test commands without specifying additional arguments.

### CODE REVIEW

**Comment by validatedev** [comment]
I have read the CLA Document and I sign the CLA

**Comment by glenn-jocher** [comment]
@validatedev thanks for the PR! We do have a `split` argument here that can be used to use test splits, i.e.:
https://github.com/ultralytics/ultralytics/blob/2193ca1658007013c25cb427b1ddfbdfeb3c3829/ultralytics/yolo/cfg/default.yaml#L43

```
yolo val split=test
```

I'm not sure if we want to define an entirely new mode, as modes are high level and we currently only have a small number of them (including mode=val), i.e. see https://docs.ultralytics.com/modes/

**Comment by validatedev** [comment]
@glenn-jocher thanks for the comment! I was the one who implemented the split feature in https://github.com/ultralytics/ultralytics/pull/704 and in the same PR you said it makes sense more when it is separated on comment https://github.com/ultralytics/ultralytics/pull/704#issuecomment-1409367229. This PR is the reflection of that request. Also in YOLOv5, dataloader indicates whether the task is val or test, so actually it makes sense for feature parity. Furthermore, it is not a completely new mode (does not propose any new logic) though and does bring really small maintenance overhead.

**Comment by glenn-jocher** [comment]
@validatedev ah ok got it!

@AyushExel @Laughing-q what do you guys think of adding a new 'test' mode, i.e. `yolo test data='coco.yaml'` in addition to the current `yolo val data='coco.yaml' split=test`

**Comment by validatedev** [comment]
Thank you, @Laughing-q, for your feedback. I understand your point. When I was working on the implementation of the `split` feature, I also had similar thoughts. However, @glenn-jocher's response about it being a useful tool if `test` is separated made me reconsider and think along the same lines. Thank you, @Laughing-q, for your feedback. I understand your point. When I was working on the implementation of the `split` feature, I also had similar thoughts. However, @glenn-jocher's response about it being a useful tool if `test` is separated made me reconsider and think along the same lines. But anyway, the final decision is yours 😊

**Comment by validatedev** [comment]
Have you reached a final decision, or do you think it's best to wait a little longer and think about the PR more before deciding? @glenn-jocher @Laughing-q @AyushExel

**Comment by glenn-jocher** [comment]
Hi @validatedev, we appreciate your interest and contribution to the YOLOv8 repository. The feature you proposed to add a separate mode `test` is something that has been recently discussed, and we are in the process of weighing the consequences of such a change. We take into account both the need for user convenience in testing their models as well as the possibility of introducing redundancies and confusion in the codebase. We kindly ask for your patience while we deliberate further on this PR and come to a final decision.

**Comment by validatedev** [comment]
@glenn-jocher, thanks for the response. I am waiting for the final decision then.

**Comment by glenn-jocher** [comment]
@validatedev thank you for your interest in YOLOv8. We appreciate your patience as we continue to evaluate the potential benefits and drawbacks of creating a separate mode for testing. Our team is diligently reviewing the PR, its implementation details, and how it aligns with the goals of the YOLOv8 repository. We will keep you updated regarding the final decision. Thank you for your understanding.

**Comment by validatedev** [comment]
@glenn-jocher been 3 months, any final decision for the PR? I can resolve the conflicts if you want also.

**Comment by validatedev** [comment]
@glenn-jocher @Laughing-q @AyushExel @RizwanMunawar could you have a chance to give an update about the final decision please? I still think it would be a great feature as an user can test their model against their test dataset easily, without any confusion.

**Comment by glenn-jocher** [comment]
Hello @validatedev, 

Thank you for your patience - we know it has been a while.
We have considered your suggestion carefully and thoroughly. We understand and appreciate the value of your feature proposal concerning ease of use for model testing. Your point about removing potential confusion for the users is well taken. 

However, in order to maintain a lean architecture and straightforward usage, we are emphasizing simplicity in the codebase of YOLOv8. Introducing a new mode such as 'test', although not a radical change, still adds another layer of complexity. While reusing existing infrastructure, i.e., the 'val' mode, we allow users to perform the same tasks without introducing a new mode. 

As a result, we have decided to maintain the current structure, which utilizes 'val' for testing with a different dataset split. 

We appreciate your understanding and thank you for your contribution to the YOLOv8 repository. We encourage you to continue to provide valuable suggestions and improvements to the project.

**Comment by validatedev** [comment]
@glenn-jocher Thank you for considering this. I understand and respect the decision. I'm happy to be part of this project and will continue to share my ideas and make improvements! Closing the PR and issue.

**Comment by glenn-jocher** [comment]
@validatedev, your understanding and contribution to the project are greatly valued. It's the enthusiastic and cooperative effort of contributors like yourself that help in the continuous improvement of YOLOv8 and its community. Please feel free to share any future ideas or enhancements you may have. Looking forward to your continued support and contributions. Happy coding!
